src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs

```diff
@@ -72,18 +72,14 @@ private static IArgumentConstraint CreateEqualityConstraint(object expressionVal
             return new EqualityArgumentConstraint(expressionValue);
         }
 
-        private static object InvokeExpression(Expression expression)
-        {
-            return Expression.Lambda(expression).Compile().DynamicInvoke();
-        }
-
         private IArgumentConstraint GetArgumentConstraintFromExpression(Expression expression)
         {
             object expressionValue = null;
 
             var trappedConstraints = this.argumentConstraintTrapper.TrapConstraints(() =>
             {
-                expressionValue = InvokeExpression(expression);
+                // in this ParsedArgumentExpression scenario, we get benefits from the same optimizations as ParsedCallExpression
+                expressionValue = ParsedCallExpression.FastEvaluate(expression);

```

Comment:
Ah, I think I see what you mean. It's just that it's still not completely clear in my mind what kind of expression is going to be fed to `FastEvaluate`. I'm going to re-read your detailed explanation in the comments :wink:

By the way, `FastEvaluate` should probably be moved out of `ParsedCallExpression` if you're going to use it from other places.